### PR TITLE
executable(): New keyword argument "link_args_pre"

### PR DESCRIPTION
**Pull request by fabio-porcedda**
Add a new keyword argument to the "executable" function.

The content of this argument will be put befere the "$in $LINK_ARGS" arguments was:
    ... {output_args} $in $LINK_ARGS ...
now is:
    ... {output_args} $LINK_ARGS_PRE $in $LINK_ARGS ...
    
Example for supporting a circular dependency:
    
    executable('name', src, link_args_pre : '-Wl,--start-group', objects : obj, link_args : '-Wl,--end-group')

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -2094,6 +2094,8 @@ def generate_link(self, target, outfile, outname, obj_list, linker, extra_args=[
             crstr = '_CROSS'
         linker_rule = linker_base + crstr + '_LINKER'
 
+        elem.add_item('LINK_ARGS_PRE', target.link_args_pre)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [F821] undefined name 'elem'

**Comment by jpakkane** [comment]
Could you please explain in more detail what sort of circular dependency you have and how does this fix it? What is in `obj`? What specifically depends on what? Is the original project public and visible on Github or elsewhere?

**Comment by fabio-porcedda** [comment]
obj is a list of binary static libs (*.a) provided without source code.
Those binary files provide functions that i use in my source code but also they require some functions that I provide in my source code.

I'm sorry but nor the source code nor the binary are public available.

**Comment by nirbheek** [comment]
We've actually been planning to use `--start-group`/`--end-group` unconditionally since this is only needed (and only available) with GNU ld on Linux. All other linkers handle circular dependencies just fine. I think that is a better solution since it solves our 'dependency explosion' problems too (#1057, #671).

**Comment by fabio-porcedda** [comment]
The downside of using --start-group/--end-group unconditionally is a performance penalty, but i don't know how much big is it.

**Comment by jpakkane** [comment]
If you have a bunch of source and a bunch of libraries like that and compile the source into obj files and then link those with the libs, the linker should make it work because obj files are not discarded after they have been linked (unlike static libraries which are).

**Comment by nirbheek** [comment]
The way the GNU linker does symbol resolution means you have to either specify the objects multiple times or use `--start-group`/`--end-group`. The linker will go over the list of objects only once, so it will simultaneously create a list of undefined references and mark them off when it encounters the definitions.

This means, `abc.o` defines A and B, but B requires C (which is in `def.o`), and C requires A (from `abc.o`), you have to specify the list of objects like this: `abc.o def.o abc.o` or use `--start-group`/`--end-group` which will go over the list of objects repeatedly till it finds all references.

**Comment by jpakkane** [comment]
That is not correct. The linker will discard symbols from static libraries but _not_ from objects. They are always looked for, even for static libraries that come after them on the linker command line. [Here is a sample repository demonstrating this](https://github.com/jpakkane/depstuff). If this were true we would need to put objects file in link targets many times for every target always (because cross using symbols within one target is what almost every target does).

**Comment by nirbheek** [comment]
So in the end it's not clear what the actual problem is. @fabio-porcedda, even though you can't tell us what the proprietary objects are, could you please at least paste the error message you get?

**Comment by fabio-porcedda** [comment]
I will check further the problem.

**Comment by fabio-porcedda** [comment]
After further investigations i found another way to fix the build issue.

I have some source files that depends on some binaries libraries, those binaries needs some function that aren't defined in my source files (only now i found that they were undefined).
Solution A: If i use the "--start-group/--end-group" my linking problems are gone, 
Solution B: i compile also some source files that defines those missing symbols without using "--start-group/--end-group"

I think that solution A works because the part of the binary libraries that use those missing symbols isn't actually needed.

Right now i'm unable to reproduce the problem using some example like the one you provided.
Right now I'm not really sure what is the correct solution.

But right now because I don't need anymore these feature i will close the pull request.

Thanks for reviewing.

**Pull request closed without merging**